File: src/plugins/lookup.ts
import { IMatrixClient, Plugin, CommandDef, MessageContext } from "./base";
import { getDb } from "../db";
import logger from "../utils/logger";
import { RateLimitsPlugin } from "./ratelimits";

const LANG_CODES = ["pt", "es", "fr", "de", "ja", "ko", "zh", "ar", "ru", "it"];

export class LookupPlugin extends Plugin {
  private rateLimitPlugin: RateLimitsPlugin;

  constructor(client: IMatrixClient, rateLimitPlugin: RateLimitsPlugin) {
    super(client);
    this.rateLimitPlugin = rateLimitPlugin;
  }

  get name() {
    return "lookup";
  }

  get commands(): CommandDef[] {
    return [
      { name: "wiki", description: "Look up a Wikipedia article", usage: "!wiki <topic>" },
      { name: "define", description: "Look up a word definition", usage: "!define <word>" },
      { name: "translate", description: "Translate text using LibreTranslate", usage: "!translate [lang] <text>" },
      { name: "urban", description: "Urban Dictionary lookup", usage: "!urban <term>" },
    ];
  }

  async onMessage(ctx: MessageContext): Promise<void> {
    if (this.isCommand(ctx.body, "wiki")) {
      await this.handleWiki(ctx);
    } else if (this.isCommand(ctx.body, "define")) {
      await this.handleDefine(ctx);
    } else if (this.isCommand(ctx.body, "translate")) {
      await this.handleTranslate(ctx);
    } else if (this.isCommand(ctx.body, "urban")) {
      await this.handleUrban(ctx);
    }
  }

  private async handleWiki(ctx: MessageContext): Promise<void> {
    const topic = this.getArgs(ctx.body, "wiki");
    if (!topic) {
      await this.sendReply(ctx.roomId, ctx.eventId, "Usage: !wiki <topic>");
      return;
    }

    try {
      const url = `https://en.wikipedia.org/api/rest_v1/page/summary/${encodeURIComponent(topic)}`;
      const res = await fetch(url);

      if (res.status === 404) {
        await this.sendReply(ctx.roomId, ctx.eventId, "No Wikipedia article found for that topic.");
        return;
      }

      if (!res.ok) {
        await this.sendReply(ctx.roomId, ctx.eventId, "Failed to fetch Wikipedia article.");
        return;
      }

      const data = await res.json() as any;
      const title = data.title ?? topic;
      let extract = data.extract ?? "";
      if (extract.length > 500) {
        extract = extract.slice(0, 500) + "...";
      }
      const pageUrl = data.content_urls?.desktop?.page ?? "";

      const reply = `${title}\n\n${extract}\n\nRead more: ${pageUrl}`;
      await this.sendMessage(ctx.roomId, reply);
    } catch (err) {
      logger.error(`Wiki lookup failed: ${err}`);
      await this.sendReply(ctx.roomId, ctx.eventId, "An error occurred while looking up that topic.");
    }
  }

  private async handleDefine(ctx: MessageContext): Promise<void> {
    const word = this.getArgs(ctx.body, "define");
    if (!word) {
      await this.sendReply(ctx.roomId, ctx.eventId, "Usage: !define <word>");
      return;
    }

    try {
      const url = `https://api.dictionaryapi.dev/api/v2/entries/en/${encodeURIComponent(word)}`;
      const res = await fetch(url);

      if (!res.ok) {
        await this.sendReply(ctx.roomId, ctx.eventId, "No definition found.");
        return;
      }

      const data = await res.json() as any[];
      if (!data || data.length === 0) {
        await this.sendReply(ctx.roomId, ctx.eventId, "No definition found.");
        return;
      }

      const entry = data[0];
      const entryWord = entry.word ?? word;
      const phonetic = entry.phonetic ?? "";
      const meaning = entry.meanings?.[0];
      const partOfSpeech = meaning?.partOfSpeech ?? "";
      const definition = meaning?.definitions?.[0]?.definition ?? "";
      const example = meaning?.definitions?.[0]?.example;

      let reply = `${entryWord}${phonetic ? ` (${phonetic})` : ""}\n[${partOfSpeech}] ${definition}`;
      if (example) {
        reply += `\nExample: "${example}"`;
      }

      await this.sendMessage(ctx.roomId, reply);
    } catch (err) {
      logger.error(`Define lookup failed: ${err}`);
      await this.sendReply(ctx.roomId, ctx.eventId, "An error occurred while looking up that word.");
    }
  }

  private async handleUrban(ctx: MessageContext): Promise<void> {
    const term = this.getArgs(ctx.body, "urban");
    if (!term) {
      await this.sendReply(ctx.roomId, ctx.eventId, "Usage: !urban <term>");
      return;
    }

    const db = getDb();
    const termLower = term.toLowerCase();
    const CACHE_TTL = 24 * 60 * 60; // 24 hours

    // Check cache
    const cached = db.prepare(
      `SELECT data FROM urban_cache WHERE term = ? AND cached_at > unixepoch() - ?`
    ).get(termLower, CACHE_TTL) as { data: string } | undefined;

    let entry: any;

    if (cached) {
      entry = JSON.parse(cached.data);
    } else {
      try {
        const url = `https://api.urbandictionary.com/v0/define?term=${encodeURIComponent(term)}`;
        const res = await fetch(url);
        if (!res.ok) {
          await this.sendReply(ctx.roomId, ctx.eventId, "Failed to look up that term.");
          return;
        }

        const data = (await res.json()) as { list: any[] };
        if (!data.list || data.list.length === 0) {
          await this.sendReply(ctx.roomId, ctx.eventId, "No Urban Dictionary entry found.");
          return;
        }

        entry = data.list[0];

        db.prepare(
          `INSERT INTO urban_cache (term, data) VALUES (?, ?)
           ON CONFLICT(term) DO UPDATE SET data = excluded.data, cached_at = unixepoch()`
        ).run(termLower, JSON.stringify(entry));
      } catch (err) {
        logger.error(`Urban Dictionary lookup failed: ${err}`);
        await this.sendReply(ctx.roomId, ctx.eventId, "An error occurred while looking up that term.");
        return;
      }
    }

    const word = entry.word ?? term;
    let definition = (entry.definition ?? "").replace(/\[|\]/g, "");
    if (definition.length > 400) definition = definition.slice(0, 400) + "...";
    let example = (entry.example ?? "").replace(/\[|\]/g, "");
    if (example.length > 200) example = example.slice(0, 200) + "...";
    const thumbsUp = entry.thumbs_up ?? 0;
    const thumbsDown = entry.thumbs_down ?? 0;

    let reply = `${word}\n\n${definition}`;
    if (example) reply += `\n\nExample: "${example}"`;
    reply += `\n\n\uD83D\uDC4D ${thumbsUp} \uD83D\uDC4E ${thumbsDown}`;

    await this.sendMessage(ctx.roomId, reply);
  }

  private async handleTranslate(ctx: MessageContext): Promise<void> {
    const LIBRETRANSLATE_URL = process.env.LIBRETRANSLATE_URL;
    if (!LIBRETRANSLATE_URL) {
      await this.sendReply(ctx.roomId, ctx.eventId, "LibreTranslate is not configured.");
      return;
    }

    const args = this.getArgs(ctx.body, "translate");
    if (!args) {
      await this.sendReply(ctx.roomId, ctx.eventId, "Usage: !translate [lang] <text>");
      return;
    }

    const allowed = this.rateLimitPlugin.checkLimit(
      ctx.sender,
      "translate",
      parseInt(process.env.RATELIMIT_TRANSLATE ?? "20"),
    );
    if (!allowed) {
      await this.sendReply(ctx.roomId, ctx.eventId, "You have reached your daily translate quota. Try again tomorrow.");
      return;
    }

    let targetLang = "en";
    let text = args;

    const words = args.split(/\s+/);
    if (words.length > 1 && words[0].length === 2 && LANG_CODES.includes(words[0].toLowerCase())) {
      targetLang = words[0].toLowerCase();
      text = words.slice(1).join(" ");
    }

    try {
      const res = await fetch(`${LIBRETRANSLATE_URL}/translate`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          q: text,
          source: "auto",
          target: targetLang,
        }),
      });

      if (!res.ok) {
        await this.sendReply(ctx.roomId, ctx.eventId, "Translation request failed.");
        return;
      }

      const data = await res.json() as any;
      const translatedText = data.translatedText ?? "";
      const detectedLang = data.detectedLanguage?.language ?? "auto";

      const reply = `[${detectedLang} \u2192 ${targetLang}] ${translatedText}`;
      await this.sendMessage(ctx.roomId, reply);
    } catch (err) {
      logger.error(`Translate failed: ${err}`);
      await this.sendReply(ctx.roomId, ctx.eventId, "An error occurred while translating.");
    }
  }
}

File: internal/plugin/lookup.go
package plugin

import (
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"net/url"
	"os"
	"strings"
	"time"

	"gogobee/internal/db"

	"maunium.net/go/mautrix"
)

var httpClient = &http.Client{Timeout: 15 * time.Second}

// LookupPlugin provides Wikipedia, dictionary, Urban Dictionary, and translation lookups.
type LookupPlugin struct {
	Base
	rateLimiter *RateLimitsPlugin
}

// NewLookupPlugin creates a new lookup plugin.
func NewLookupPlugin(client *mautrix.Client, rateLimiter *RateLimitsPlugin) *LookupPlugin {
	return &LookupPlugin{
		Base:        NewBase(client),
		rateLimiter: rateLimiter,
	}
}

func (p *LookupPlugin) Name() string { return "lookup" }

func (p *LookupPlugin) Commands() []CommandDef {
	return []CommandDef{
		{Name: "wiki", Description: "Look up a Wikipedia summary", Usage: "!wiki <topic>", Category: "Lookup & Reference"},
		{Name: "define", Description: "Look up a word definition", Usage: "!define <word>", Category: "Lookup & Reference"},
		{Name: "urban", Description: "Look up Urban Dictionary definition", Usage: "!urban <term>", Category: "Lookup & Reference"},
		{Name: "translate", Description: "Translate text (pt/es/fr/de/ja/ko/zh/ar/ru/it)", Usage: "!translate [lang] <text>", Category: "Lookup & Reference"},
	}
}

func (p *LookupPlugin) Init() error { return nil }

func (p *LookupPlugin) OnReaction(_ ReactionContext) error { return nil }

func (p *LookupPlugin) OnMessage(ctx MessageContext) error {
	switch {
	case p.IsCommand(ctx.Body, "wiki"):
		return p.handleWiki(ctx)
	case p.IsCommand(ctx.Body, "define"):
		return p.handleDefine(ctx)
	case p.IsCommand(ctx.Body, "urban"):
		return p.handleUrban(ctx)
	case p.IsCommand(ctx.Body, "translate"):
		return p.handleTranslate(ctx)
	}
	return nil
}

func (p *LookupPlugin) handleWiki(ctx MessageContext) error {
	topic := strings.TrimSpace(p.GetArgs(ctx.Body, "wiki"))
	if topic == "" {
		return p.SendReply(ctx.RoomID, ctx.EventID, "Usage: !wiki <topic>")
	}

	encoded := url.PathEscape(strings.ReplaceAll(topic, " ", "_"))
	apiURL := fmt.Sprintf("https://en.wikipedia.org/api/rest_v1/page/summary/%s", encoded)

	resp, err := httpClient.Get(apiURL)
	if err != nil {
		slog.Error("lookup: wiki request", "err", err)
		return p.SendReply(ctx.RoomID, ctx.EventID, "Failed to reach Wikipedia.")
	}
	defer resp.Body.Close()

	if resp.StatusCode == 404 {
		return p.SendReply(ctx.RoomID, ctx.EventID, fmt.Sprintf("No Wikipedia article found for \"%s\".", topic))
	}
	if resp.StatusCode != 200 {
		return p.SendReply(ctx.RoomID, ctx.EventID, "Wikipedia returned an error.")
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return p.SendReply(ctx.RoomID, ctx.EventID, "Failed to read Wikipedia response.")
	}

	var result struct {
		Title       string `json:"title"`
		Extract     string `json:"extract"`
		ContentURLs struct {
			Desktop struct {
				Page string `json:"page"`
			} `json:"desktop"`
		} `json:"content_urls"`
	}
	if err := json.Unmarshal(body, &result); err != nil {
		return p.SendReply(ctx.RoomID, ctx.EventID, "Failed to parse Wikipedia response.")
	}

	extract := result.Extract
	if len(extract) > 500 {
		extract = extract[:500] + "..."
	}

	msg := fmt.Sprintf("%s\n\n%s\n\n%s", result.Title, extract, result.ContentURLs.Desktop.Page)
	return p.SendReply(ctx.RoomID, ctx.EventID, msg)
}

func (p *LookupPlugin) handleDefine(ctx MessageContext) error {
	word := strings.TrimSpace(p.GetArgs(ctx.Body, "define"))
	if word == "" {
		return p.SendReply(ctx.RoomID, ctx.EventID, "Usage: !define <word>")
	}

	apiURL := fmt.Sprintf("https://api.dictionaryapi.dev/api/v2/entries/en/%s", url.PathEscape(word))

	resp, err := httpClient.Get(apiURL)
	if err != nil {
		slog.Error("lookup: define request", "err", err)
		return p.SendReply(ctx.RoomID, ctx.EventID, "Failed to reach dictionary API.")
	}
	defer resp.Body.Close()

	if resp.StatusCode == 404 {
		return p.SendReply(ctx.RoomID, ctx.EventID, fmt.Sprintf("No definition found for \"%s\".", word))
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return p.SendReply(ctx.RoomID, ctx.EventID, "Failed to read dictionary response.")
	}

	var entries []struct {
		Word     string `json:"word"`
		Meanings []struct {
			PartOfSpeech string `json:"partOfSpeech"`
			Definitions  []struct {
				Definition string `json:"definition"`
				Example    string `json:"example"`
			} `json:"definitions"`
		} `json:"meanings"`
	}

	if err := json.Unmarshal(body, &entries); err != nil || len(entries) == 0 {
		return p.SendReply(ctx.RoomID, ctx.EventID, fmt.Sprintf("No definition found for \"%s\".", word))
	}

	entry := entries[0]
	var sb strings.Builder
	sb.WriteString(fmt.Sprintf("Definition of \"%s\":\n\n", entry.Word))

	for i, meaning := range entry.Meanings {
		if i >= 3 { // Limit to 3 meanings
			break
		}
		sb.WriteString(fmt.Sprintf("(%s)\n", meaning.PartOfSpeech))
		for j, def := range meaning.Definitions {
			if j >= 2 { // Limit to 2 definitions per meaning
				break
			}
			sb.WriteString(fmt.Sprintf("  %d. %s\n", j+1, def.Definition))
			if def.Example != "" {
				sb.WriteString(fmt.Sprintf("     Example: \"%s\"\n", def.Example))
			}
		}
	}

	return p.SendReply(ctx.RoomID, ctx.EventID, sb.String())
}

func (p *LookupPlugin) handleUrban(ctx MessageContext) error {
	term := strings.TrimSpace(p.GetArgs(ctx.Body, "urban"))
	if term == "" {
		return p.SendReply(ctx.RoomID, ctx.EventID, "Usage: !urban <term>")
	}

	d := db.Get()
	termLower := strings.ToLower(term)

	// Check cache (24h)
	var cachedData string
	var cachedAt int64
	err := d.QueryRow(
		`SELECT data, cached_at FROM urban_cache WHERE term = ?`,
		termLower,
	).Scan(&cachedData, &cachedAt)

	now := time.Now().UTC().Unix()
	if err == nil && now-cachedAt < 86400 {
		return p.SendReply(ctx.RoomID, ctx.EventID, cachedData)
	}

	// Fetch from API
	apiURL := fmt.Sprintf("https://api.urbandictionary.com/v0/define?term=%s", url.QueryEscape(term))

	resp, err := httpClient.Get(apiURL)
	if err != nil {
		slog.Error("lookup: urban request", "err", err)
		return p.SendReply(ctx.RoomID, ctx.EventID, "Failed to reach Urban Dictionary.")
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return p.SendReply(ctx.RoomID, ctx.EventID, "Failed to read Urban Dictionary response.")
	}

	var result struct {
		List []struct {
			Word       string `json:"word"`
			Definition string `json:"definition"`
			Example    string `json:"example"`
			ThumbsUp   int    `json:"thumbs_up"`
			ThumbsDown int    `json:"thumbs_down"`
		} `json:"list"`
	}

	if err := json.Unmarshal(body, &result); err != nil || len(result.List) == 0 {
		return p.SendReply(ctx.RoomID, ctx.EventID, fmt.Sprintf("No Urban Dictionary definition found for \"%s\".", term))
	}

	entry := result.List[0]
	def := entry.Definition
	// Remove bracket notation used by Urban Dictionary
	def = strings.ReplaceAll(def, "[", "")
	def = strings.ReplaceAll(def, "]", "")
	if len(def) > 400 {
		def = def[:400] + "..."
	}

	example := entry.Example
	example = strings.ReplaceAll(example, "[", "")
	example = strings.ReplaceAll(example, "]", "")
	if len(example) > 200 {
		example = example[:200] + "..."
	}

	var sb strings.Builder
	sb.WriteString(fmt.Sprintf("Urban Dictionary: %s\n\n", entry.Word))
	sb.WriteString(fmt.Sprintf("%s\n", def))
	if example != "" {
		sb.WriteString(fmt.Sprintf("\nExample: %s\n", example))
	}
	sb.WriteString(fmt.Sprintf("\n+%d / -%d", entry.ThumbsUp, entry.ThumbsDown))

	msg := sb.String()

	// Cache the result
	_, err = d.Exec(
		`INSERT INTO urban_cache (term, data, cached_at) VALUES (?, ?, ?)
		 ON CONFLICT(term) DO UPDATE SET data = ?, cached_at = ?`,
		termLower, msg, now, msg, now,
	)
	if err != nil {
		slog.Error("lookup: urban cache", "err", err)
	}

	return p.SendReply(ctx.RoomID, ctx.EventID, msg)
}

var supportedLangs = map[string]bool{
	"pt": true, "es": true, "fr": true, "de": true, "ja": true,
	"ko": true, "zh": true, "ar": true, "ru": true, "it": true,
}

func (p *LookupPlugin) handleTranslate(ctx MessageContext) error {
	args := strings.TrimSpace(p.GetArgs(ctx.Body, "translate"))
	if args == "" {
		return p.SendReply(ctx.RoomID, ctx.EventID,
			"Usage: !translate [lang] <text>\nSupported: pt, es, fr, de, ja, ko, zh, ar, ru, it")
	}

	ltURL := os.Getenv("LIBRETRANSLATE_URL")
	if ltURL == "" {
		return p.SendReply(ctx.RoomID, ctx.EventID, "Translation service is not configured.")
	}

	// Rate limit: 10 translations per day
	if p.rateLimiter != nil && !p.rateLimiter.CheckLimit(ctx.Sender, "translate", 10) {
		remaining := p.rateLimiter.Remaining(ctx.Sender, "translate", 10)
		return p.SendReply(ctx.RoomID, ctx.EventID,
			fmt.Sprintf("Translation rate limit reached. %d remaining today.", remaining))
	}

	// Parse lang code and text
	parts := strings.SplitN(args, " ", 2)
	targetLang := "es" // default
	text := args

	if len(parts) >= 2 && supportedLangs[strings.ToLower(parts[0])] {
		targetLang = strings.ToLower(parts[0])
		text = parts[1]
	}

	// Call LibreTranslate
	payload := fmt.Sprintf(`{"q":%q,"source":"auto","target":%q}`, text, targetLang)
	req, err := http.NewRequest("POST", strings.TrimRight(ltURL, "/")+"/translate", strings.NewReader(payload))
	if err != nil {
		return p.SendReply(ctx.RoomID, ctx.EventID, "Failed to create translation request.")
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := httpClient.Do(req)
	if err != nil {
		slog.Error("lookup: translate request", "err", err)
		return p.SendReply(ctx.RoomID, ctx.EventID, "Failed to reach translation service.")
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return p.SendReply(ctx.RoomID, ctx.EventID, "Failed to read translation response.")
	}

	var result struct {
		TranslatedText string `json:"translatedText"`
		Error          string `json:"error"`
	}
	if err := json.Unmarshal(body, &result); err != nil {
		return p.SendReply(ctx.RoomID, ctx.EventID, "Failed to parse translation response.")
	}

	if result.Error != "" {
		return p.SendReply(ctx.RoomID, ctx.EventID, fmt.Sprintf("Translation error: %s", result.Error))
	}

	msg := fmt.Sprintf("Translation (-> %s):\n%s", targetLang, result.TranslatedText)
	return p.SendReply(ctx.RoomID, ctx.EventID, msg)
}
